src/serac/coefficients/coefficient_extensions.hpp

```diff
@@ -14,41 +14,170 @@
 
 #include <functional>
 #include <memory>
-
+#include <variant>
+#include <vector>
 #include "mfem.hpp"
 
 #include "serac/numerics/expr_template_ops.hpp"
-
+#include "serac/physics/utilities/finite_element_state.hpp"
 /**
  * @brief Functionality that extends current MFEM capabilities
  *
  */
 namespace serac::mfem_ext {
 
+namespace detail {
+// methods for determining index type
+template <typename T>
+struct index_t {
+  using type = std::size_t;
+};
+
+template <typename T>
+struct index_t<mfem::Array<T>> {
+  using type = int;
+};
+
+// methods for determining the size of a container
+template <typename T>
+auto size(T container)
+{
+  return container.size();
+}
+
+template <typename T>
+std::size_t size(mfem::Array<T> container)
+{
+  return container.Size();
+}
+
+// methods for determining type of coefficient evaluations
+template <typename T, typename = void>
+struct eval_t {
+  using type = T;
+};
+
+template <typename T>
+struct eval_t<T, std::enable_if_t<std::is_base_of_v<mfem::Coefficient, T>>> {
+  using type = double;
+};
+
+template <typename T>
+struct eval_t<T, std::enable_if_t<std::is_base_of_v<mfem::VectorCoefficient, T>>> {
+  using type = mfem::Vector;
+};
+
+/// methods for evaluating coefficient stuff with the same prototype
+double eval(double& d, mfem::ElementTransformation&, const mfem::IntegrationPoint&) { return d; }
+
+/// evaluates an mfem::Coefficient
+double eval(mfem::Coefficient& c, mfem::ElementTransformation& Tr, const mfem::IntegrationPoint& ip)
+{
+  return c.Eval(Tr, ip);
+}
+
+/// evaluates a mfem::VectorCoefficient
+mfem::Vector eval(mfem::VectorCoefficient& v, mfem::ElementTransformation& Tr, const mfem::IntegrationPoint& ip)
+{
+  mfem::Vector temp(v.GetVDim());
+  v.Eval(temp, Tr, ip);
+  return temp;
+}
+
+
+template <typename T, typename = void>
+struct is_iterable : std::false_type {};
+  
+template <typename T>
+struct is_iterable<T, std::void_t<decltype(std::declval<T>().begin()),
+                                  decltype(std::declval<T>().end())>>
+    : std::true_type {};
+
+template <typename T, typename = void>
+struct has_brackets : std::false_type {};
+  
+template <typename T>
+struct has_brackets<T, std::void_t<decltype(std::declval<T>()[0])>>
+    : std::true_type {};
+  
+}  // namespace detail
+
+namespace digitize {
```

Comment:
I am not super sold on this name for the namespace.  Also these seem more like general utility functions that should go in a common place like `infrastructure`

Reply:
Good point. Currently it's only used in conjunction with coefficient_extensions, but they have mostly been provided for convenience. All of the functions take in a `double` and return an `int`.  Let me know if there's a better namespace.